File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main (void)
{
    long number = get_long("number: ");
    int length = 0;
    long buffer1 = number;
    long buffer2 = 0;
    long buffer3 = 0;
    long add = 0;
    long check = 0;

    while(buffer1 > 0) {
        buffer1 /= 10;
        length++;
    }

    buffer1 = number;

    for(int i = 0; i < length; i++) {
        buffer1 /= 10;
        buffer2 = buffer1 % 10;
        buffer2 *= 2;
        if (buffer2 > 9)
        {
            buffer3 = buffer2 % 10;
            buffer2 = buffer3 + (buffer2/10);
        }
        add += buffer2;
        buffer1 /= 10;
    }

    buffer1 = number;

    for(int i = 0; i < length; i++) {
        buffer2 = buffer1 % 10;
        buffer1 /= 100;
        add += buffer2;
    }

    if (add % 10 != 0) {
        printf("Invalid\n");
        return 1;
    }

    check = number / pow(10,length-1);

    switch (check) {
        case 4:
            if (length == 13 || length == 16) {
                printf("VISA\n");
            }
            break;
    }

    check = number / pow(10,length-2);

    switch (check) {
        case 37:
        case 34:
            if (length == 15) {
                printf("American Express\n");
            }
            break;

        case 51:
        case 52:
        case 53:
        case 54:
        case 55:
            if (length == 16) {
                printf("MasterCard\n");
            }
            break;
    }
    return 0;
}

File: pset6/python/credit/credit.py
from cs50 import get_int

def main():

    number = get_int("Nummer: ")
    length = len(str(number))
    buffer1 = number
    addnumber = 0

    for i in range(length):
        buffer1 = buffer1 // 10
        buffer2 = buffer1 % 10
        buffer2 = buffer2 * 2
        if buffer2 > 9:
            buffer3 = buffer2 % 10
            buffer2 = buffer3 + (buffer2//10)
        addnumber = addnumber + buffer2
        buffer1 = buffer1 // 10

    buffer1 = number

    for i in range(length):
        buffer2 = buffer1 % 10
        buffer1 = buffer1 // 100
        addnumber = addnumber + buffer2

    check = number // pow(10,length-1)

    if check == 4:
        if length == 13 or length == 16:
            print("VISA")

    check = number // pow(10,length-2)

    if check == 34:
        if length == 15:
            print("American Express")
    elif check == 37:
        if length == 15:
            print("American Express")
    elif check == 51:
        if length == 16:
            print("MasterCard")
    elif check ==52:
        if length == 16:
            print("MasterCard")
    elif check == 53:
        if length == 16:
            print("MasterCard")
    elif check == 54:
        if length == 16:
            print("MasterCard")
    elif check == 55:
        if length == 16:
            print("MasterCard")

    if addnumber % 10 == 0:
        print("Valid")
    else:
        print("Invalid")

if __name__ == "__main__":
    main()